### PR TITLE
Move "global" internal slots to MediaDevices, and fix "relevant object" links

### LINKED ISSUE #938: Global object relevant to what?
This spec says things like _"the [relevant global object](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global)"_ (16 hits) without specifying which platform object it is relevant _to_.

For [mediaDevices](https://w3c.github.io/mediacapture-main/getusermedia.html#mediadevices) this is particularly noticeable as we create internal slots _"On page load, ... On the [relevant global object](https://html.spec.whatwg.org/multipage/#concept-relevant-global)"_, which seems wrong. The internal slots should probably be on the mediaDevices object itself, and initalized on its creation, not on _"page load"_.

### PR DESCRIPTION
**Pull request by jan-ivar**
Fixes #938. Using [hide whitespace](https://github.com/w3c/mediacapture-main/pull/939/files?diff=unified&w=1) when reviewing is recommended.

***
<a href="https://pr-preview.s3.amazonaws.com/jan-ivar/mediacapture-main/pull/939.html" title="Last updated on Mar 22, 2023, 8:26 PM UTC (6894336)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/mediacapture-main/939/92ea82d...jan-ivar:6894336.html" title="Last updated on Mar 22, 2023, 8:26 PM UTC (6894336)">Diff</a>

### CODE REVIEW

**Comment by jan-ivar** [comment]
In cleaning this up, I came across a couple of realities that I've tried to handle better with this PR.

1. There's more than one mediaDevices object to consider. E.g. [in iframes](https://jan-ivar.github.io/dummy/iframe_iframe_gum_cross.html).
2. A source only stops when _ALL_ tracks in the entire user agent have ended, so [stop all sources](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-stop-all-sources) on ONE global object using [[[mediaStreamTrackSources]]](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-mediastreamtracksources) doesn't work. I reworked it to lean on existing [ends](https://w3c.github.io/mediacapture-main/getusermedia.html#ends-nostop) algorithms instead, which seems to work well even for transfered tracks.
3. Some of the internal slots were (and still are) accessed off main thread
4. For [Privacy Indicator Requirements](https://w3c.github.io/mediacapture-main/getusermedia.html#privacy-indicator-requirements) I just added a note saying dealing with multiple iframes is an exercise left to UAs.

**Comment by youennf** [comment]
> 2\. A source only stops when _ALL_ its tracks across the entire user agent have ended

A UA can stop a source and all tracks will get ended.
This happens when for instance the camera is unplugged.

>, so [stop all sources](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-stop-all-sources) on ONE global object using [[[mediaStreamTrackSources]]](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-mediastreamtracksources) doesn't work.

I do not see why it would not work, getUserMedia creates a source that gets associated to ONE global object.
It is true that in the current spec, ending the tracks is needed to not trigger needless ended events at unload time.
For transferred tracks, stopping the source will trigger the ended event, which does not seem bad.

**Comment by jan-ivar** [comment]
> I do not see why it would not work, getUserMedia creates a source that gets associated to ONE global object.

My understanding is a [source](https://w3c.github.io/mediacapture-main/#dfn-source) _"can be a physical webcam"_, making it a singleton in the entire user agent.

[gUM](https://w3c.github.io/mediacapture-main/#dfn-same-permission) doesn't "create" it, as much as it: _"Let[s] grantedDevice be finalCandidate's source device. ... Let track be the result of [creating a MediaStreamTrack](https://w3c.github.io/mediacapture-main/#dfn-create-a-mediastreamtrack) with grantedDevice"_, which initializes track.[[[Source]]](https://w3c.github.io/mediacapture-main/#dfn-source-0).

E.g. in [this iframe example](https://jan-ivar.github.io/dummy/iframe_iframe_gum_cross.html) (Once I hit Allow and the `Camera` button in the iframe) two video element self-views will play, but there is ONE source (the singular physical camera) associated with TWO global objects (two `navigator.mediaDevices` objects with distinct [relevant global object](https://html.spec.whatwg.org/multipage/webappapis.html#concept-relevant-global)s).

> For transferred tracks, stopping the source will trigger the ended event

But [stop all sources](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-stop-all-sources) isn't handling that correctly, as written. It says: _"For each [MediaStreamTrack](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamtrack) object track whose [relevant global object](https://html.spec.whatwg.org/multipage/#concept-relevant-global) is globalObject, set track's [[[ReadyState]]](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-readystate) to "[ended](https://w3c.github.io/mediacapture-main/getusermedia.html#idl-def-MediaStreamTrackState.ended)"."_

This fails to:
1. queue a task
2. consider the same source may be tied to multiple global objects
3. include tracks that have been transferred away to other global objects that now should end (since [track source lifetime management remains tied to the context that created it](https://w3c.github.io/mediacapture-extensions/#transferable-mediastreamtrack))
4. consider tracks that have been transferred _here_ from _other_ global objects (which get included here), but if this was the last reference, it needs to [end](https://www.w3.org/TR/mediacapture-streams/#ends-nostop) that reference and stop the source in those other global objects)

In the PR this is fixed [like this](https://pr-preview.s3.amazonaws.com/jan-ivar/mediacapture-main/pull/939.html#dfn-end-all-mediastreamtracks).

**Comment by jan-ivar** [comment]
> A UA can stop a source and all tracks will get ended.
> This happens when for instance the camera is unplugged.

That sounds like [ends for any reason other than the stop() method being invoked](https://w3c.github.io/mediacapture-main/#ends-nostop), not [stop all sources](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-stop-all-sources) of a [global object](https://html.spec.whatwg.org/multipage/webappapis.html#global-object). Unplugging a camera should stop ONE source, ending all tracks of that source regardless of how many global objects it is associated with.

**Comment by youennf** [comment]
> My understanding is a [source](https://w3c.github.io/mediacapture-main/#dfn-source) _"can be a physical webcam"_, making it a singleton in the entire user agent.

Ah, this is where we are not aligned.
I do not think the spec works well with source as a singleton.
For instance, it is useful to be able to have all tracks of a document tracks being muted while tracks of another document tracks are not muted, even if they both are using the same camera. Safari is using this for instance to have only one tab capturing at a time. If the source was a singleton, it would not be possible.

AIUI, the main (only) requirement is that cloned tracks have to share the same source.
This is the minimum granularity that all UAs adhere to and this is what we can use as a basis.

> This fails to:
> 
> 1. queue a task

We should dig into this, stopping the sources happen when the document gets unloaded, it is not clear whether queueing a task at this time will ensure the task will be executed.

> 2. consider the same source may be tied to multiple global objects

It works as follows:
- the source has tracks in different global objects
- only one global object will trigger stopping the source when global object gets unloaded
- all tracks that are not in the source global object will get ended after the source is stopped (with an ended event).

> 3. include tracks that have been transferred away to other global objects that now should end (since [track source lifetime management remains tied to the context that created it](https://w3c.github.io/mediacapture-extensions/#transferable-mediastreamtrack))

See above, this is handled.

> 4. consider tracks that have been transferred _here_ from _other_ global objects (which get included here), but if this was the last reference, it needs to [end](https://www.w3.org/TR/mediacapture-streams/#ends-nostop) that reference and stop the source in those other global objects)

As per above, I am not sure why this would be needed.

**Comment by jan-ivar** [comment]
> I do not think the spec works well with source as a singleton.
> For instance, it is useful to be able to have all tracks of a document tracks being muted while tracks of another document tracks are not muted, even if they both are using the same camera. Safari is using this for instance to have only one tab capturing at a time. If the source was a singleton, it would not be possible.

As you say, _tracks_ are [muted](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamtrack-muted), not sources, so I don't immediately see what's not possible. But let's dig in:

[§ 4.3.1.2 Media Flow](https://w3c.github.io/mediacapture-main/getusermedia.html#media-flow) begins with _"[Muted](https://w3c.github.io/mediacapture-main/getusermedia.html#track-muted) refers to the input to the [MediaStreamTrack](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamtrack)."_, BUT this ultimately just describes the [track's muted state](https://w3c.github.io/mediacapture-main/getusermedia.html#set-track-muted), and when the UA should update it, which it accomplishes (almost!) without mentioning "source".

The section has just two mentions of "source" near the end:
1. _"media from the source only flows when a [MediaStreamTrack](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamtrack) object is both unmuted and enabled."_, which is fine
2. _"For a newly created [MediaStreamTrack](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediastreamtrack) object, the following applies: the track is always enabled unless stated otherwise (for example when cloned) and the muted state reflects the state of the source at the time the track is created."_

The second one does mention _"the muted state reflects the state of the source"_, which seems a tacit admission that the source has some (unnamed) state. But I think s/source/input/ here would have been more self-consistent with the rest of this section.

I think what happened here was that all the user agent examples we could think of at the time were global to the user agent (_"the user pushing a physical mute button on the microphone, the user closing a laptop lid with an embedded camera, the user toggling a control in the operating system"_), and we wanted to express the invariant that creation of new tracks shouldn't sidestep those user-directed privacy measures, so we had no choice but to tacitly refer to some state outside of the track.¹

I think the challenge here is how to express this invariant while at the same time allowing a user agent to mute all tracks in some contexts and not others. — But I don't think defining multiple instances of _source_ is the answer, because having an instance live inside a single (per realm) global object seems too limiting, because in my view, the idea that gUM for the same camera from a page and from an iframe in that page (see my example [above](https://github.com/w3c/mediacapture-main/pull/939#issuecomment-1458870504)) refers to distinct source objects seems confusing for little benefit. At least in Firefox, privacy indicators act as if this is the same source (yes I gave up fixing that section in the spec).

> AIUI, the main (only) requirement is that cloned tracks have to share the same source. This is the minimum granularity that all UAs adhere to and this is what we can use as a basis.

[Transferred](https://w3c.github.io/mediacapture-extensions/#transferable-mediastreamtrack) tracks also share the same [[Source]] of their [detached](https://html.spec.whatwg.org/multipage/structured-data.html#detached) counterpart. Such tracks may be transferred not just to workers but to other documents, in which case their relevant global object (which will be in the destination realm) won't have it in its [[[mediaStreamTrackSources]]](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-mediastreamtracksources). 

This, iframes, and the way privacy indicators have been implemented in all browsers, I think suggest the minimal granularity is a page. E.g. if an iframe calls gUM, transfers the resulting video track up to the document containing it, and then the iframe navigates, does the the video track end? If so, why should it? The permissions were delegated to the iframe in the first place, so why can capture not continue? So perhaps the source should live in the top-level document? Or maybe we can write algorithms that don't depend on where sources are defined? 

> > This fails to:
> > 
> > 1. queue a task
> 
> We should dig into this, stopping the sources happen when the document gets unloaded, it is not clear whether queueing a task at this time will ensure the task will be executed.

Good point. I can remove that.

> It works as follows:
> 
>     * the source has tracks in different global objects
> 
>     * only one global object will trigger stopping the source when global object gets unloaded
> 
>     * all tracks that are not in the source global object will get ended after the source is stopped (with an ended event).

My problem with this and the current text is it only works correctly if we ASSUME sources are different per global object, which at best isn't obvious and at worst seems to lack support in the spec, which means when I read this (thinking sources are singletons) it reads to me as doing the WRONG thing (stopping other concurrent sources). In contrast, I hope my PR does not rely on that assumption (since it was written without it).

---
<sub>1. There's also the more recent: _"[[[Muted]]](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-muted), initialized to true if source is [muted](https://w3c.github.io/mediacapture-main/getusermedia.html#track-muted)"_, which I think erroneously refers to the track muted state.</sub>

**Comment by youennf** [comment]
> My problem with this and the current text is it only works correctly if we ASSUME sources are different per global object, which at best isn't obvious

I think the intent of the spec is clear, as we discussed this when designing transferred streams.
We can always clarify this in the spec.

**Comment by jan-ivar** [comment]
I see my mistake now. I misread what the [original text](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-stop-all-sources) was trying to do (I read it as inadvertently stopping all tracks of a physical source in the user agent). But in fixing it, it fails to preserve the original intent, which I failed to interpret and we wish to to preserve, which is effectively to end all tracks "tied" to the `mediaDevices` object they came from (through `mediaDevices.getUserMedia()`), so that even tracks transferred elsewhere end when this document unloads. Or as mediacapture-extensions [says](https://w3c.github.io/mediacapture-extensions/#transferable-mediastreamtrack): _"When originalContext goes away, trackSource gets ended, thus transferredTrack gets ended."_

I'll try to update my PR to not break that. To be fair, I believe the existing language only accomplishes that under ONE interpretation of the scope of source (one per global object), NOT the other (source is a singleton of the user agent), but I believe we can express while being agnostic of source scope.

> I think the intent of the spec is clear, as we discussed this when designing transferred streams.
We can always clarify this in the spec.

This is all I'm trying to do. :)

**Comment by youennf** [comment]
> To be fair, I believe the existing language only accomplishes that under ONE interpretation of the scope of source (one per global object), NOT the other (source is a singleton of the user agent)

The assumption is more one source per getUserMedia call (well per track created by getUserMedia).
This mirrors well the principle of one source/capture controller per getDisplayMedia call.
Could we get a consensus on this model from an editorial point of view?

**Comment by jan-ivar** [comment]
> The assumption is more one source per getUserMedia call (well per track created by getUserMedia).

Where in the spec is such an assumption relied on?

Rather, the [source stopped](https://w3c.github.io/mediacapture-main/getusermedia.html#source-stopped) algorithm relies on a 1-1 between a source and device:
- _"When all tracks using a source have been stopped or ended by some other means, the source is stopped. If the source is a device exposed by [getUserMedia](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediadevices-getusermedia)(), then when the source is stopped, the UA MUST run the following steps:_

1. _Let deviceId be the device's [deviceId](https://w3c.github.io/mediacapture-main/getusermedia.html#dom-mediadeviceinfo-deviceid)._
2. _Set [[[devicesLiveMap]]](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-deviceslivemap)[deviceId] to false."_

If JS calls gUM twice and get two video tracks with the same deviceId, and then I end one of them, a per-track source interpretation would turn off the privacy indicator for this realm even though the other video track is still live.

The [[[devicesLiveMap]]](https://w3c.github.io/mediacapture-main/getusermedia.html#dfn-deviceslivemap) is on the global object (which is 1-1 with `navigator.mediaDevices`), 

That seems like one too many 1-to-many relationships.

**Comment by jan-ivar** [comment]
I've simplified this PR to focus on the move of internal slots, which should hopefully be editorial now. @youennf PTAL.

File: getusermedia.html [hunk 1]
```diff
@@ -870,12 +871,13 @@ <h4>Life-cycle</h4>
...
           <li>
-            <p>Notify <var>track</var>'s source that <var>track</var> is
+            <p>Inform <var>track</var>'s {{MediaStreamTrack/[[Source]]}} that <var>track</var> is
             [= track/ended =] so that the source may be [= source/stopped =], unless other
             {{MediaStreamTrack}} objects depend on it.</p>
           </li>
           <li>
-            <p>[=Fire an event=] named <a data-link-type=event>ended</a> at the object.</p>
+            <p>If <var>notify</var> is `true`, [=fire an event=] named <a data-link-type=event>ended</a>
+            at <var>track</var>.</p>
```
Comment (jan-ivar) [inline getusermedia.html:880]: I missed this one. Will remove it before merging.

File: getusermedia.html [hunk 2]
```diff
@@ -2766,24 +2793,26 @@ <h3>{{MediaDevices}}</h3>
       unavailable, or the system default for input and/or output devices of a
       {{MediaDeviceKind}} changed, the [=User Agent=] MUST run the following
       <dfn>device change notification steps</dfn> for each {{MediaDevices}}
-      object for which [=device enumeration can proceed=] is <code>true</code>,
+      object, <var>mediaDevices</var>, for which [=device enumeration can proceed=] is <code>true</code>,
       but for no other {{MediaDevices}} object:</p>
       <ol>
+          <li>
+            <p>Let <var>document</var> be <var>mediaDevices</var>'s
```
[the remark this answers is not in the document]
Comment (jan-ivar) [reply getusermedia.html]: Thanks for spotting this leftover unused variable!

**Review by youennf — APPROVED** [review]

**Comment by jan-ivar** [comment]
The PR now should be true to its title and only move internal slots that were on the global object (`window`) to the `navigator.mediaDevices` object. There should be no behavioral difference from this.

**Pull request merged**